Critical trio.BusyResourceError when joining a channel

Steps to reproduce:

* First user is matterbot (matterbridge bot) he is connected in a channel
* Second user is DrLazor (hexchat user), he joins the channel
* Critical error

Tracekback:

	[IO] recv from DrLazor: b'JOIN #second-salon'
	[DEBUG] Dispatch to JOIN: ('#second-salon',)
	[IO] send to #second-salon: :DrLazor JOIN #second-salon
	[IO] send to DrLazor: :DrLazor JOIN #second-salon
	[IO] send to matterbot: :DrLazor JOIN #second-salon
	[IO] recv from matterbot: b'WHO DrLazor %tacuhnr,1'
	[DEBUG] Dispatch to WHO: ('DrLazor', '%tacuhnr,1')
	[WARNING] Command WHO sent by matterbot failed, code: 400
	[IO] send to matterbot: :drlazor.be 400 matterbot - :Command WHO is unknown.
	[ERROR] Error while serving matterbot.
	Traceback (most recent call last):
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/user.py", line 144, in serve
	    await self.state.dispatch(*args)
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/states.py", line 42, in dispatch
	    raise ErrUnknownError(self.user, '-', f"Command {cmd} is unknown.")
	aioircd.exceptions.ErrUnknownError: :drlazor.be 400 matterbot - :Command WHO is unknown.
	During handling of the above exception, another exception occurred:
	Traceback (most recent call last):
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/server.py", line 28, in handle
	    await user.serve()
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/user.py", line 148, in serve
	    await self.send(exc.args[0])
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/user.py", line 172, in send
	    await self.stream.send_all(b"".join(f"{msg}\r\n".encode() for msg in messages))
	  File "/usr/local/lib/python3.9/dist-packages/trio/_highlevel_socket.py", line 99, in send_all
	    with self._send_conflict_detector:
	  File "/usr/local/lib/python3.9/dist-packages/trio/_util.py", line 190, in __enter__
	    raise trio.BusyResourceError(self._msg)
	trio.BusyResourceError: another task is currently sending data on this SocketStream
	[INFO] Terminate connection of matterbot
	[IO] send to #général: :matterbot QUIT :[':Internal', 'host', 'error']
	[IO] send to #second-salon: :matterbot QUIT :[':Internal', 'host', 'error']
	[IO] send to DrLazor: :drlazor.be 353 DrLazor = #second-salon :DrLazor
	[IO] send to DrLazor: :matterbot QUIT :[':Internal', 'host', 'error']
	[IO] send to DrLazor: :drlazor.be 366 DrLazor #second-salon :End of /NAMES list.
	[CRITICAL] Dead
	Traceback (most recent call last):
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/__main__.py", line 43, in main
	    trio.run(server.serve)
	  File "/usr/local/lib/python3.9/dist-packages/trio/_core/_run.py", line 1932, in run
	    raise runner.main_task_outcome.error
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/server.py", line 52, in serve
	    await trio.serve_tcp(self.handle, self.port, host=self.addr)
	  File "/usr/local/lib/python3.9/dist-packages/trio/_highlevel_open_tcp_listeners.py", line 219, in serve_tcp
	    await trio.serve_listeners(
	  File "/usr/local/lib/python3.9/dist-packages/trio/_highlevel_serve_listeners.py", line 121, in serve_listeners
	    task_status.started(listeners)
	  File "/usr/local/lib/python3.9/dist-packages/trio/_core/_run.py", line 815, in __aexit__
	    raise combined_error_from_nursery
	  File "/usr/local/lib/python3.9/dist-packages/trio/_highlevel_serve_listeners.py", line 25, in _run_handler
	    await handler(stream)
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/server.py", line 34, in handle
	    await user.terminate("Internal host error")
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/user.py", line 158, in terminate
	    await self.state.QUIT(f":{kick_msg}".split(' '), kick=True)
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/states.py", line 104, in QUIT
	    await chan.send(f":{self.user.nick} QUIT :{reason}")
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/channel.py", line 35, in send
	    self._nursery.start_soon(partial(
	  File "/usr/local/lib/python3.9/dist-packages/trio/_core/_run.py", line 815, in __aexit__
	    raise combined_error_from_nursery
	  File "/usr/local/lib/python3.9/dist-packages/aioircd/user.py", line 172, in send
	    await self.stream.send_all(b"".join(f"{msg}\r\n".encode() for msg in messages))
	  File "/usr/local/lib/python3.9/dist-packages/trio/_highlevel_socket.py", line 99, in send_all
	    with self._send_conflict_detector:
	  File "/usr/local/lib/python3.9/dist-packages/trio/_util.py", line 190, in __enter__
	    raise trio.BusyResourceError(self._msg)
	trio.BusyResourceError: another task is currently sending data on this SocketStream

--- a/aioircd/user.py
+++ b/aioircd/user.py
@@ -50,6 +50,7 @@ class User:
         self.state = (PasswordState if servlocal.pwd else ConnectedState)(self)
         self.channels = set()
         self._ping_timer = trio.CancelScope()  # dummy
+        self._send_lock = trio.StrictFIFOLock()
 
     def __str__(self):
         if self.nick:
@@ -166,7 +167,8 @@ class User:
         if isinstance(messages, str):
             messages = [messages]
 
-        if log:
-            for msg in messages:
-                logger.log(aioircd.IO, "send to %s: %s", self, msg)
-        await self.stream.send_all(b"".join(f"{msg}\r\n".encode() for msg in messages))
+        async with self._send_lock:
+            if log:
+                for msg in messages:
+                    logger.log(aioircd.IO, "send to %s: %s", self, msg)
+            await self.stream.send_all(b"".join(f"{msg}\r\n".encode() for msg in messages))

